src/main/java/org/codehaus/mojo/flatten/KeepCommentsInPom.java

```diff
@@ -0,0 +1,236 @@
+package org.codehaus.mojo.flatten;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.logging.Log;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.w3c.dom.bootstrap.DOMImplementationRegistry;
+import org.w3c.dom.ls.DOMImplementationLS;
+import org.w3c.dom.ls.LSOutput;
+import org.w3c.dom.ls.LSSerializer;
+import org.xml.sax.SAXException;
+
+/**
+ * Helper class to keep the comments how they have been in the original pom.xml
+ * While reading with {@link org.apache.maven.model.io.xpp3.MavenXpp3Writer} the comments are not placed into the {@link org.apache.maven.model.Model}
+ * and so {@link org.apache.maven.model.io.xpp3.MavenXpp3Writer} is not able to re-write those comments.
+ * 
+ * Workaround (maybe until core is fixed) is to remember all the comments and restore them after MavenXpp3Writer has created the new flattened pom.xml.
+ * 
+ * Current restriction on non-unique child nodes is that this class finds the node back due to the position in the file,
+ * that may lead to mis-re-added comments e.g. on multiple added dependencies
+ * (but for e.g. resolveCiFriendliesOnly the nodes keep stable)
+ * 
+ */
+class KeepCommentsInPom {
+
+	/**
+	 * Create an instance with collected current comments of the passed pom.xml file.
+	 */
+	static KeepCommentsInPom create(Log aLog, File aOriginalPomFile) throws MojoExecutionException {
+		KeepCommentsInPom tempKeepCommentsInPom = new KeepCommentsInPom();
+		tempKeepCommentsInPom.setLog(aLog);
+		tempKeepCommentsInPom.loadComments(aOriginalPomFile);
+		return tempKeepCommentsInPom;
+	}
+
+	private Log log;
+	
+	/**
+	 * The unique path list for an original node (the comments are stored via the referenced previousSibling)
+	 */
+	private Map<String,AtomicReference<Node>> commentsPaths;

```

Comment:
Why AtomicReference? I think this is used from a single thread only.

Reply:
You are right, AtomicReference is more for threading. 
When taking java.util.Optional as replacement it would not really be optional (as BiConsumer would never get an empty one). I will check.

Reply:
> So I take Optional.

Value of Optional cannot be set later, so I need to take something other.

Reply:
I changed it to pass the nodePath to the BiConsumer directly.